File: csharp/modules/Numbers.Utils/Numbers.Utils.Tests/Operations.Tests/Factorial.Utils.Tests/FactorialUtilsTests.cs
using System.Numerics;

namespace Prog.Numbers.Utils.Operations.Tests;

[TestFixture]
public class FactorialUtilsTests
{
    /// <summary> <h1><c>FormatErrorMessage</c></h1>
    /// <br/><br/>This method is used to format the error message.<br/><br/>
    ///
    /// <b>Arguments:</b>
    ///
    /// <list type="bullet">
    /// <item>
    /// <c>expected: BigInteger</c> - The expected list of factors.
    /// </item>
    /// <item>
    /// <c>actual: BigInteger</c> - The actual list of factors.
    /// </item>
    /// </list>
    ///
    /// <b>Return:</b>
    ///
    /// <list type="bullet">
    /// <item>
    /// <c>string</c> - The formatted error message.
    /// </item>
    /// </list>
    /// </summary>
    private static string FormatErrorMessage(BigInteger expected, BigInteger actual)
    {
        return $"\nTest run at: {DateTime.Now:O}\nExpected : {expected}\nActual   : {actual}\n";
    }

    /// <summary> <h1><c>FactorialTest</c></h1>
    /// <br/><br/>This method is used to test the Factorial method of
    /// FactorialUtils.<br/><br/>
    /// </summary>
    [Test]
    public void FactorialTest()
    {
        int n = 75;
        BigInteger expectedOutput = BigInteger.Parse(
            "24809140811395398091946477116594033660926243886570122837795894512655842677572867409443815424000000000000000000"
        );
        BigInteger actualOutput = FactorialUtils.Factorial(n);
        string errorMessage = FormatErrorMessage(expectedOutput, actualOutput);
        Assert.That(expectedOutput, Is.EqualTo(actualOutput), errorMessage);
    }

    /// <summary> <h1><c>FactorialRandomTest</c></h1>
    /// <br/><br/>This method is used to test the FactorialRandom method of
    /// FactorialUtils.<br/><br/>
    /// </summary>
    [Test]
    public void FactorialRandomTest()
    {
        string result = FactorialUtils.FactorialRandom();
        string startString = "Fact[";
        Assert.That(
            result,
            Does.StartWith(startString),
            $"\nExpected: {startString}\n     Got: {result}\n"
        );
        Assert.That(result, Does.Contain(" = "));
    }
}

File: haskell/packages/numbers-utils/tests/Operations/FactorialUtilsTests.hs
{-# LANGUAGE OverloadedStrings #-}

module Operations.FactorialUtilsTests (factorialTest, factorialRandomTest) where

import Data.List (isInfixOf, isPrefixOf)

import Data.Time.Clock (getCurrentTime)
import Data.Time.LocalTime (utcToLocalZonedTime)
import Test.Tasty (TestTree)
import Test.Tasty.HUnit (assertBool, assertEqual, testCase)

import Numbers.Utils.Operations.FactorialUtils (factorial, factorialRandom)

-- |
-- # @formatErrorMessage@
--
-- The function @formatErrorMessage@ is used to format the error message.
--
-- __Arguments:__
--
-- * @expected: String@ - The expected output.
-- * @actual: String@ - The actual output.
--
-- __Return:__
--
-- * @IO String@ - The formatted error message.
formatErrorMessage :: String -> String -> IO String
formatErrorMessage expected actual = do
    now <- getCurrentTime >>= utcToLocalZonedTime
    return $
        "\nTest run at: "
            ++ show now
            ++ "\nExpected : "
            ++ expected
            ++ "\nActual   : "
            ++ actual
            ++ "\n"

-- |
-- # @factorialTest@
--
-- The function @factorialTest@ is used to test the @factorial@ function.
factorialTest :: TestTree
factorialTest = testCase "Factorial" $ do
    let num = 75
    let expectedOutput =
            read
                "24809140811395398091946477116594033660926243886570122837795894512655842677572867409443815424000000000000000000"
                :: Integer
    let actualOutput = factorial num
    errorMessage <- formatErrorMessage (show expectedOutput) (show actualOutput)
    assertEqual errorMessage expectedOutput actualOutput

-- |
-- # @factorialRandomTest@
--
-- The function @factorialRandomTest@ is used to test the @factorialRandom@
-- function.
factorialRandomTest :: TestTree
factorialRandomTest = testCase "Factorial Random Test" $ do
    result <- factorialRandom
    let startString = "Fact["
    let errorMessage = "\nExpected: " ++ startString ++ "\n     Got: " ++ result ++ "\n"
    assertBool errorMessage (startString `isPrefixOf` result)
    assertBool errorMessage (" = " `isInfixOf` result)
